File: src/experiments/tetris/rl/encode.ts
import { COLS, ROWS, type GameState, type PieceType } from '../engine'
import {
  PIECE_TYPES,
  RL_OBS_CHANNELS,
  RL_OBS_COLS,
  RL_OBS_ROWS,
  type RlModelMetadata,
} from './contracts'

export function pieceTypeIndex(type: PieceType): number {
  return PIECE_TYPES.indexOf(type)
}

/** Encode game state into channel-major float32 tensor [C, H, W]. */
export function encodeObservation(
  state: GameState,
  gravityNorm: number,
  out?: Float32Array,
): Float32Array {
  const size = RL_OBS_CHANNELS * RL_OBS_ROWS * RL_OBS_COLS
  const tensor = out ?? new Float32Array(size)
  tensor.fill(0)

  const stride = RL_OBS_ROWS * RL_OBS_COLS

  for (let r = 0; r < ROWS; r++) {
    for (let c = 0; c < COLS; c++) {
      const idx = r * COLS + c
      const boardVal = state.board[r]![c]!
      if (boardVal > 0) tensor[idx] = boardVal / 7
    }
  }

  if (state.piece) {
    const piece = state.piece
    for (let r = 0; r < piece.matrix.length; r++) {
      for (let c = 0; c < piece.matrix[r]!.length; c++) {
        if (!piece.matrix[r]![c]) continue
        const x = piece.x + c
        const y = piece.y + r
        if (y < 0 || y >= ROWS || x < 0 || x >= COLS) continue
        const idx = y * COLS + x
        tensor[stride + idx] = 1
      }
    }

    const rotNorm = piece.rot / 3
    const fyNorm = Math.max(0, Math.min(1, piece.fy))
    const chRot = 2 * stride
    const chFy = 3 * stride
    for (let i = 0; i < stride; i++) {
      tensor[chRot + i] = rotNorm
      tensor[chFy + i] = fyNorm
    }
  }

  const nextIdx = pieceTypeIndex(state.next)
  if (nextIdx >= 0) {
    const ch = (4 + nextIdx) * stride
    for (let i = 0; i < stride; i++) tensor[ch + i] = 1
  }

  const g = Math.max(0, Math.min(1, gravityNorm))
  const chGravity = 11 * stride
  for (let i = 0; i < stride; i++) tensor[chGravity + i] = g

  return tensor
}

export function boardFingerprint(state: GameState): string {
  const parts: string[] = []
  for (let r = 0; r < ROWS; r++) {
    parts.push(state.board[r]!.join(''))
  }
  if (state.piece) {
    const p = state.piece
    parts.push(
      `P:${p.type},${p.rot},${p.x},${p.y},${p.fy.toFixed(3)}`,
    )
  } else {
    parts.push('P:null')
  }
  parts.push(`N:${state.next}`)
  parts.push(`S:${state.score},L:${state.lines},GO:${state.gameOver ? 1 : 0}`)
  return parts.join('|')
}

export function defaultMetadata(partial?: Partial<RlModelMetadata>): RlModelMetadata {
  return {
    version: 1,
    actionNames: [
      'noop',
      'left',
      'right',
      'rotateCW',
      'rotateCCW',
      'softDrop',
      'hardDrop',
    ],
    decisionIntervalMs: 100,
    obsChannels: RL_OBS_CHANNELS,
    obsRows: RL_OBS_ROWS,
    obsCols: RL_OBS_COLS,
    gravityMin: 1,
    gravityMax: 6,
    ...partial,
  }
}

export function normalizeGravity(cellsPerSec: number, meta: RlModelMetadata): number {
  const span = Math.max(1e-6, meta.gravityMax - meta.gravityMin)
  return Math.max(0, Math.min(1, (cellsPerSec - meta.gravityMin) / span))
}

File: tetris-rl/tetris_rl/encode.py
"""Observation encoding — mirrors src/experiments/tetris/rl/encode.ts."""

from __future__ import annotations

import numpy as np

from .engine import COLS, ROWS, GameState, PieceType

PIECE_TYPES: list[PieceType] = ["I", "O", "T", "S", "Z", "J", "L"]
RL_OBS_CHANNELS = 12
RL_OBS_ROWS = 20
RL_OBS_COLS = 10


def piece_type_index(piece_type: PieceType) -> int:
    return PIECE_TYPES.index(piece_type)


def encode_observation(state: GameState, gravity_norm: float) -> np.ndarray:
    tensor = np.zeros((RL_OBS_CHANNELS, RL_OBS_ROWS, RL_OBS_COLS), dtype=np.float32)
    stride = RL_OBS_ROWS * RL_OBS_COLS

    for r in range(ROWS):
        for c in range(COLS):
            val = state.board[r][c]
            if val > 0:
                tensor[0, r, c] = val / 7.0

    if state.piece:
        piece = state.piece
        for r, row in enumerate(piece.matrix):
            for c, val in enumerate(row):
                if not val:
                    continue
                x = piece.x + c
                y = piece.y + r
                if y < 0 or y >= ROWS or x < 0 or x >= COLS:
                    continue
                tensor[1, y, x] = 1.0

        rot_norm = piece.rot / 3.0
        fy_norm = max(0.0, min(1.0, piece.fy))
        tensor[2, :, :] = rot_norm
        tensor[3, :, :] = fy_norm

    next_idx = piece_type_index(state.next)
    if next_idx >= 0:
        tensor[4 + next_idx, :, :] = 1.0

    g = max(0.0, min(1.0, gravity_norm))
    tensor[11, :, :] = g

    return tensor
